poppy/sub_sampled_optics.py

```diff
@@ -0,0 +1,253 @@
+                     detector=None,
+                     overwrite_inputwavefront=False,
+                     display_intermediates=False,
+                     optical_system=None,
+                 **kwargs):
+            self.n_apertures = dimensions[0]*dimensions[1]
+            
+            self.optic_array=optic_array
+            self.crosstalk=crosstalk
+            if crosstalk:
+                raise ValueError("CROSS TALK NOT IMPLEMENTED <YET>")
+            self.x_y_offset=x_y_offset
+            self.amplitude = np.asarray([1.])
+            self.opd = np.asarray([0.])
+            self.pixelscale = None
+            self.input_wavefront = None
+            self.output_wavefront = None
+            if detector == None:
+                self.detector = Detector(0.01,fov_pixels=128)
+            else:
+                self.detector=detector
+            self.optical_system = optical_system
+            if  optical_system is not None:
+                    raise ValueError("complete optical system after wavelets are not implemented yet")
+            self.x_apertures=self.optic_array.shape[0]
+            self.y_apertures=self.optic_array.shape[1]
+            if self.x_apertures != self.y_apertures:
+                raise ValueError("A square array of subapertures is currently required.")
+            #initialize array of subsampled output wavefronts:
+            self.wf_array = np.empty(self.optic_array.shape,dtype=np.object_)
+            self.overwrite_inputwavefront = overwrite_inputwavefront
+            self.display_intermediates = display_intermediates
+            self._propagated_flag = False #can't have propagated when initializing
+            OpticalElement.__init__(self, **kwargs)
+                
+        def sample_wf(self, wf):
+            '''
+            
+            Parameters
+            ----------
+            
+            '''
+            #save the input wavefront
+            self.input_wavefront = wf
+            for i in range(self.x_apertures):
+                 for j in  range(self.y_apertures):
+                    opt = self.optic_array[i][j] #get an optic
+
+                    #check for padding
+                    if opt == None:
+                        continue
+                    
+                    aper_per_dim = wf.diam /(opt.pupil_diam) #assuming squares
+                    
+                    self._w = opt.pupil_diam/wf.pixelscale #subaperture width in pixels 
+                    #the generated number of subapertures might not match the input wavefront dimensions
+                    #want to center the subapertures on the incoming wavefront
+                    
+                    self.c = wf.wavefront.shape[0]/2*u.pix #center of array
+                    c=self.c
+                    w=self._w
+                    sub_wf=wf.copy() #new wavefront has all the previous wavefront properties
+                    lower_x = int((c + w*(i)  - w*self.x_apertures/2).value)
```

Comment:
This block of repeated code is a good candidate for factoring out into a function or static method. That will minimize the chance that an arithmetic error fixed in one place doesn't get fixed in all three.